readme.md
# FullyLinked 
## A simple library for visualising graphs using DOM elements



### 🎉 Features

- ✅ Written with [TypeScript](https://www.typescriptlang.org/) so it is type-safe
- ✅ Draggable nodes
- ✅ Add edges dynamically with drag & drop
- ✅ Zoomable & draggable canvas
- ✅ Nodes can be any HTML elements (e.g. `<div>` or `<img>`) or even a React component
- ✅ Add nodes dynamically with drag & drop
- ✅ Events:
  - 🎉 `nodeClick`: when a node is clicked
  - 🎉 `nodeDrag`: when a node is dragged
  - 🎉 `nodeDragEnd`: when a node is dropped
  - 🎉 `edgeClick`: when an edge is clicked
  - 🎉 `canvasClick`: when the canvas is clicked
  - 🎉 `canvasDrag`: when the canvas is dragged
  - 🎉 `canvasDragEnd`: when the canvas is dropped
  - 🎉 `canvasZoom`: when the canvas is zoomed
  - 🎉 `canvasZoomEnd`: when the canvas is zoomed
  - and many more [see FullyLinkedEventEnum](src/common/event/FullyLinkedEventEnum.ts)
- ✅ Extend with graph layouts, including DAG, force-directed, or a custom layout (coming soon)
- ✅ Customizable edge styles (coming soon)


### 🔀 Example FullyLinked Graph
![Example FullyLinked Graph](src/example/simple/example.png)

### ℹ️ Get started

1. First, install the library

```bash
npm install fully-linked
```

2. Then import it
```typescript
import { FullyLinked } from 'fully-linked'
```

3. Use it
```typescript

// Create an interface for our node data
interface MyNodeDataType {
  id: string;
  label: string;
}

// Create an interface for our edge data
interface MyEdgeDataType {
  id: string;
  label: string;
}


// Get the container element. This can be any HTML div element. 
// It must have a width and height so that the graph can be visible.
const container = document.getElementById("container");
if (!container) {
    // ... handle container not found
}

// Create a new FullLinked instance
const fl = new FullyLinked<MyNodeDataType, MyEdgeDataType>({
  // Each graph must have a unique ID
  id: "my-fl-graph",
  // The container element must be passed in
  container,
});

// A function to create your node template. 
// Returns an HTML element.
const getNodeElement = (node: ProcessedNode<MyNodeDataType>) => {
  // Create a blank div element as our node container
  const nodeElement = document.createElement("div");

  // Set the node container's styles
  nodeElement.style.width = `${node.width}px`;
  nodeElement.style.height = `${node.height}px`;
  nodeElement.style.backgroundColor = "#eeee";
  nodeElement.style.borderRadius = "10px";

  // Add node content
  const content = document.createElement('div');
  content.style.padding = "15px";
  content.innerText = node.data.label;

  // Append the node content to the node container
  nodeElement.appendChild(content);

  // Return the node container
  return nodeElement;
};

// Set graph data
fl.setData({
  nodes: [
    {
      // Each node must have a unique id
      id: "1",
      // Each node must have a width and a height  
      width: 300,
      height: 80,
      // Initial positions of the node
      x: 50,
      y: 50,
      // Pass any data you want to store in the node
      data: { id: "1", label: "Node 1" },
      // The node template is a function that returns an HTML element (created above)
      customNodeElement: getNodeElement
    },
    {
      id: "2",
      width: 300,
      height: 80,
      x: 500,
      y: 250,
      data: { id: "2", label: "Node 2" },
      customNodeElement: getNodeElement
    },
    {
      id: "3",
      width: 300,
      height: 80,
      x: 50,
      y: 450,
      data: { id: "3", label: "Node 3" },
      customNodeElement: getNodeElement
    },
    {
      id: "4",
      width: 300,
      height: 80,
      x: 50,
      y: 650,
      data: { id: "4", label: "Node 4" },
      customNodeElement: getNodeElement
    },
    {
      id: "5",
      width: 300,
      height: 80,
      x: 50,
      y: 850,
      data: { id: "5", label: "Node 5" },
      customNodeElement: getNodeElement
    },
    {
      id: "6",
      width: 300,
      height: 80,
      x: 50,
      y: 1000,
      data: { id: "6", label: "Node 6" },
      customNodeElement: getNodeElement
    }
  ],
  edges: [
    { 
        // An edge must have a unique ID
        id: "1", 
        // The source and target nodes of the edge
        source: "1",
        target: "2",
        // Pass any data you want to store in the edge
        data: { 
            id: "1", 
            label: "Edge 1" 
        } 
    },
    { id: "2", source: "2", target: "3", data: { id: "2", label: "Edge 2" } },
    { id: "3", source: "2", target: "4", data: { id: "3", label: "Edge 3" } },
    { id: "4", source: "4", target: "5", data: { id: "4", label: "Edge 4" } },
  ],
  // Each data object must have a unique id
  id: "test",
});


// Render the graph into the container element
fl.render();

```

CHANGELOG.md
# Changelog

## [2.2.1] - 2022-01-09
### Fixed
- Node not draggable if x and y are initially 0

## [2.2.0] - 2022-01-09
### Added
- Ability to customize edge styles
## [2.1.1] - 2022-01-05
### Fixed
- Enums not exported
- Issue with imported ReactElement Type in d.ts

## [2.1.0] - 2022-01-02
### Fixed
- globalNodeProps not passed into nodes
- Make GlobalNodeProps type required

## [2.0.1] - 2022-01-02
### Fixed
- Update readme with the new type `ProcessedNode`

## [2.0.0] - 2022-01-02
### Added
- Introduce GlobalNodeProps in `FullyLinkedOptions`, which comes with a new generic type
- Rename InternalNode to ProcessedNode to reflect the fact that it is a node processed by FullyLinked
## [1.0.61] - 2022-01-01
### Added
- Add 'isRoot' property to all nodes to indicate whether they are root nodes

## [1.0.60] - 2021-12-31
### Removed
- Removed xUnadjustedByZoom and yUnadjustedByZoom as not needed

## [1.0.59] - 2021-12-31

### Added
- Add xUnadjustedByZoom and yUnadjustedByZoom to the InternalNode type so that user can use them to save the positions of the nodes and restore the positions in a new graph instance by setting the FullyLink option `initialCamera`
## [1.0.58] - 2021-12-31
### Fixed
- Refactored how initial zoom and pan is set so that is is more readible.
## [1.0.57] - 2021-12-31
### Added
- Ability to add initial canvas zoom and transform
- Canvas zoom and transform events

## [1.0.56] - 2021-12-30
### Fixed
- Fix release issue

## [1.0.53] - 2021-12-30

### Fixed
- Node drag end event should not be emitted many times


## [1.0.52] - 2021-12-30

### Added
- Update readme to include events
## [1.0.51] - 2021-12-30
### Added
- Initial FullyLinked events
- Add pino logger library

## [1.0.49] - 2021-12-30 - 2021-12-30

### Added
- test .npmignore file
- add .npmignore file and then removed it

## [1.0.48] - 2021-12-30

### Added

- Keep-a-changelog plugin added.

## [1.0.47] - 2021-01-08
### Added

- Add test for diffItems 38d3943
- Enable auto-changelog 6d742af


src/example/simple/example.ts

import { ProcessedNode } from "../../common/item/node/types/Node";
import { FullyLinked } from "../../index";

interface MyNodeDataType {
  id: string;
  label: string;
}

interface MyEdgeDataType {
  id: string;
  label: string;
}

const fullyLinked = new FullyLinked<MyNodeDataType, MyEdgeDataType, any>({
  id: "test-graph",
  container: document.getElementById("container") as HTMLDivElement,
});

const getNodeElement = (node: ProcessedNode<MyNodeDataType, any>) => {
  const nodeElement = document.createElement("div");
  nodeElement.style.width = `${node.width}px`;
  nodeElement.style.height = `${node.height}px`;
  nodeElement.style.backgroundColor = "#eeee";
  nodeElement.style.borderRadius = "10px";
  const content = document.createElement('div');
  content.style.padding = "15px";
  content.innerText = node.data.label;
  nodeElement.appendChild(content);
  return nodeElement;
};

fullyLinked.setData({
  nodes: [
    {
      id: "1",
      width: 300,
      height: 80,
      x: 50,
      y: 50,
      data: { id: "1", label: "Node 1" },
      customNodeElement: getNodeElement,
    },
    {
      id: "2",
      width: 300,
      height: 80,
      x: 500,
      y: 250,
      data: { id: "2", label: "Node 2" },
      customNodeElement: getNodeElement,
    },
    {
      id: "3",
      width: 300,
      height: 80,
      x: 1000,
      y: 450,
      data: { id: "3", label: "Node 3" },
      customNodeElement: getNodeElement,
    },
    {
      id: "4",
      width: 300,
      height: 80,
      x: 1000,
      y: 650,
      data: { id: "4", label: "Node 4" },
      customNodeElement: getNodeElement,
    },
    {
      id: "5",
      width: 300,
      height: 80,
      x: 1600,
      y: 850,
      data: { id: "5", label: "Node 5" },
      customNodeElement: getNodeElement,
    },
    {
      id: "6",
      width: 300,
      height: 80,
      x: 50,
      y: 1000,
      data: { id: "6", label: "Node 6" },
      customNodeElement: getNodeElement,
    },
  ],
  edges: [
    { id: "1", source: "1", target: "2", data: { id: "1", label: "Edge 1" } },
    { id: "2", source: "2", target: "3", data: { id: "2", label: "Edge 2" } },
    { id: "3", source: "2", target: "4", data: { id: "3", label: "Edge 3" } },
    { id: "4", source: "4", target: "5", data: { id: "4", label: "Edge 4" } },
  ],
  id: "test",
});

fullyLinked.render();
